# An analysis on Apps Stores: most profitable apps available on march 2019

In this project we will present you some insights on the most downloaded (and the most profitable ones, as a consequence) apps on mobile app stores.

Our main objective is to help our team to understand what kinds of apps are likely to attract more users on the Google Play Store and Apple Apps Store, so we can make data driven decisions about what types of application are likely to create more revenue once developed and distributed.

For both systems there's a huge collection of apps available, about 2 million apps on the App Store and 2,1 million on Google Play Store. Such large universe would be impractical to analyse, since it would require too much time spent on data cleasing, and extra costs aquiring such dataset.

Thus, bearing in mind the statistics principle of sampling, we've opted for using two free datasets, one for each platform, still containing a good deal of apps, large enough to 

We got this number using a free dataset which is available [here](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

That dataset is organized in columns, and described as follows:

| Column | Description |
|---|---|
|"id" | App ID |
| "track_name" | App Name |
| "size_bytes" | Size (in Bytes) |
| "currency" | Currency Type |
| "price" | Price amount |
| "rating_count_tot" | User Rating counts (for all version) |
| "rating_count_ver" | User Rating counts (for current version) |
| "user_rating" | Average User Rating value (for all version) |
| "user_rating_ver" | Average User Rating value (for current version) |
| "ver" | Latest version code |
| "cont_rating" | Content Rating |
| "prime_genre" | Primary Genre |
| "sup_devices.num" | Number of supporting devices |
| "ipadSc_urls.num" | Number of screenshots showed for display |
| "lang.num" | Number of supported languages |
| "vpp_lic" | Vpp Device Based Licensing Enabled |


For Android the columns are quite self-explanatory:

| Column |
|---|
|App|
|Category|
|Rating|
|Reviews|
|Size|
|Installs|
|Type|
|Price|
|Content Rating|
|Genres|
|Last Updated|
|Current Ver|
|Android Ver|


**We will start exploring the datasets by importing a function called *reader*, from the **CSV library**, which means *comma separated values*, a general use format for importing datasheets and lists organized in columns and rows. Further, we open and load each file in a format that can be manipulated with Python commands.**

In [1]:
from csv import reader

opened_file_apple = open('AppleStore.csv')
read_apple = reader(opened_file_apple)
apple_apps_data = list(read_apple)

opened_file_android = open('googleplaystore.csv')
read_android = reader(opened_file_android)
android_apps_data = list(read_android)

**Now, we're going to import a function that organizes dictionaries (a special kind of data list) in decreasing order. Its use we'll see much  further on this work.**

In [2]:
from collections import OrderedDict

**Here, after we loaded the datasets, we print the first row in each one. Our objective is to show what information is available on those tables. For greater convenience, we also created two variables with the rows, which we will use many times for reference.**

In [3]:
android_header = (android_apps_data[0])
print (android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [4]:
apple_header = (apple_apps_data[0])
print (apple_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


**We've created a simple function for exploring the datasets, which can be seen below:**

In [5]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

**And here we use that function with the two datasets.**

In [6]:
explore_data(apple_apps_data, 1, 4, rows_and_columns=True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7198
Number of columns: 16


In [7]:
explore_data(android_apps_data, 1, 4, rows_and_columns=True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10842
Number of columns: 13


**With further analysis on these datasets we've found apps that lacked information, and redundant apps. If not rectified, those occurences would create statistical distortions and make this work less reliable.**

**So, we had to cleanse the sets, removing duplicates and items which lacked information.**

**Below, we start the data cleansing, removing from the Android list an app that lacked an important column for our analysis.**

In [8]:
index = 0
for row in android_apps_data[1:]:
    index += 1
    app_name = row[0]
    if app_name == 'Life Made WI-Fi Touchscreen Photo Frame':
        print (row)
        print (index)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10473


In [9]:
del android_apps_data[10473]

**Now we check for duplicates on the Android dataset, creating two lists, one for unique registers, and one for duplicated ones.**

In [10]:
unique_android = []
duplicated_android = []

for row in android_apps_data[1:]:
    name = row[0]
    if name in unique_android:
        duplicated_android.append(name)
    else:
        unique_android.append(name)

**We'll do the same for the iOS dataset.**

In [11]:
unique_ios = []
duplicated_ios = []

for row in apple_apps_data[1:]:
    name = row[1]
    if name in unique_ios:
        duplicated_ios.append(name)
    else:
        unique_ios.append(name)

**As we do this checking, we found 1.181 duplicated items in Android dataset, from a total of 10.840. This redundances represent more than 10% of the registers and given its weight, it could lead to wrong conclusions further.**

**For Apple dataset, we've found 2 redundances, which could be ignored without trouble for our final results.**

In [12]:
and_dupl_num = len (duplicated_android)
and_uniq_num = len (unique_android)
ios_dupl_num = len (duplicated_ios)
ios_uniq_num = len (unique_ios)

print (and_dupl_num)
print (and_uniq_num)
print (ios_dupl_num)
print (ios_uniq_num)

1181
9659
2
7195


**Here's some of these redundances we've found in Android set:**

In [13]:
print (duplicated_android[:10])

['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


**Many of these redundances have columns with different values, so we can not just randomly them. Instagram duplicated registries, for example, have different values on the 4th column, which holds the total number of reviews for a given app:**

In [14]:
for row in android_apps_data[1:]:
    name = row[0]
    if name == 'Instagram':
        print (row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


**We came to conclusion that there's a relation amongst the numbers on that columns: they have a growing pattern what indicates the growing number of reviews for that app, so we just kept the largest number (as it would indicate the latest value) and removed the smaller ones**

In [15]:
reviews_max = {}

for row in android_apps_data[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if name not in reviews_max:
        reviews_max[name] = n_reviews
    if name in reviews_max:
        if n_reviews > reviews_max[name]:
            reviews_max[name] = n_reviews

**Now we can check the "reviews_max" dictionary length, which contains no duplicated items. We expect it to match the unique values number of 9656, as seen on the first checking for duplicates**

In [16]:
len (reviews_max)

9659

**It matches the number we've found on the first checking. Now we're going to clear our main list.**

In [17]:
android_clean = []
apple_clean = apple_apps_data[1:]
already_added = []

for row in android_apps_data[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if n_reviews == reviews_max[name]:
        if name not in already_added:
            android_clean.append(row)
            already_added.append(name)
            

**Let's check if the new list matches the numbers of unique apps and explore it a bit**

In [18]:
len (android_clean)

9659

**The list length is ok. Now with the exploring:**

In [19]:
explore_data(android_clean, 1, 7)

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


['Smoke Effect Photo Maker - Smoke Editor', 'ART_AND_DESIGN', '3.8', '178', '19M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'April 26, 2018', '1.1', '4.0.3 and up']


['Infinite Painter', 'ART_AND_DESIGN', '4.1', '36815', '29M', '1,000,000+', 'Free', '0', 'Everyone', 

**Now, we're going to check the list of apps added to the clean list, just to make sure it corresponds the number of unique apps, and explore it a litte, to confirm its content.**

In [20]:
len (already_added)
print (already_added[1:6])

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint', 'Pixel Draw - Number Art Coloring Book', 'Paper flowers instructions', 'Smoke Effect Photo Maker - Smoke Editor']


**Everything seems ok. Now it's time to remove apps with non-latin characters, since our reasearch is about apps that could make a profit on Western countries.**

**This can be made by using the ASCII system codification to characters, where every character has a corresponding number, and where the latin characters range from 0 to 127**

**In Python, there's a built-in function to check a character's ASCII corresponding number, which is "ord". We're going to use this to check if a app name has non-latin characters, and if it is the case, we're remove that app from our list, as follows.**

In [21]:
def check_latin (app_name):
    latin_char = []
    non_latin_char = []
    for char in app_name:
        if ord(char) < 128:
            latin_char.append(char)
        if ord(char) > 127:
            non_latin_char.append(char)
    
    len_latin = len (latin_char)
    len_non_latin = len (non_latin_char)
    if len_latin > len_non_latin:
        return True
    else:
        return False

real_clean_googleapps = []
for row in android_clean[1:]:
    ck = check_latin(row[0])
    if ck == True:
        real_clean_googleapps.append(row)
        
real_clean_appleapps = []
for row in apple_clean[1:]:
    chk = check_latin(row[1])
    if chk == True:
        real_clean_appleapps.append(row)

**Now, we're gonna check the ultimate clean lists lengths, just to make sure everything looks right**

In [22]:
goog_len = len (real_clean_googleapps)
appl_len = len (real_clean_appleapps)
print (goog_len)
print (appl_len)

9632
6226


**A brief look at these dataset revealed that free to use apps, which rely on ads to profit, are the most numerous and downloaded ones. We created a simple program to show this.**

In [23]:
apple_apps_prices = {'Free': 0, 'Non_free': 0}
apple_apps_downloads = {'Free': 0, 'Non_free': 0}
for row in real_clean_appleapps:
    price = float (row[4])
    downloaded_times = float (row[5])
    if price == 0.0:
        apple_apps_prices['Free'] += 1
        apple_apps_downloads['Free'] += downloaded_times
    else:
        apple_apps_prices['Non_free'] += 1
        apple_apps_downloads['Non_free'] += downloaded_times

total_apps = apple_apps_prices['Free'] + apple_apps_prices['Non_free']
total_downloads = apple_apps_downloads['Free'] + apple_apps_downloads['Non_free']

**And below, you can see the numbers. In this sample, there are 4.056 free apps, downloaded 80.105.208 times, and 3.141 paid apps, downloaded 12.685.045 times.**

In [24]:
print (apple_apps_prices)

{'Free': 3253, 'Non_free': 2973}


In [25]:
print (apple_apps_downloads)

{'Free': 77011249.0, 'Non_free': 12680532.0}


**We can have a better notion in percentual terms, as we see below:**

In [26]:
apple_free_apps_propor = (apple_apps_prices['Free']/total_apps)*100
apple_free_downloads = (apple_apps_downloads['Free']/total_downloads)*100

In [27]:
proporcao_apps = round(apple_free_apps_propor)
proporcao_downloads = round (apple_free_downloads)
print (proporcao_apps)
print (proporcao_downloads)

52
86


**As we can see, free apps represent 56% of total apps and 86% of all downloads in iOS App Store.**

**Now it's time to isolate free apps on the two apps stores. Let's check again the header rows to get the indexes of prices**

In [28]:
print (android_apps_data[0])
print (apple_apps_data[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


**As we see, the indexes are [7] for Android Apps and [4] for iOS Apps. Let's check the ultimate clean lists to confirm.**

In [29]:
print (real_clean_googleapps[0])
print (real_clean_appleapps[0])

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


**Everything looks fine. Now we're going to isolate the free apps on the two datasets.**

In [30]:
android_free = []
android_non_free = []
ios_free = []
ios_non_free = []

for row in real_clean_googleapps:
    price = row[6]
    if price == 'Free':
        android_free.append(row)
    else:
        android_non_free.append(row)
    
    
for row in real_clean_appleapps:
    price = row[4]
    if price == '0.0':
        ios_free.append(row)
    else:
        ios_non_free.append(row)

**Let's explore the free datasets, starting by Android**

In [31]:
explore_data(android_free, 0, 5)

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


['Smoke Effect Photo Maker - Smoke Editor', 'ART_AND_DESIGN', '3.8', '178', '19M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'April 26, 2018', '1.1', '4.0.3 and up']




**Now, with the iOS dataset**

In [32]:
explore_data(ios_free, 0, 5)

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


['429047995', 'Pinterest', '74778624', 'USD', '0.0', '1061624', '1814', '4.5', '4.0', '6.26', '12+', 'Social Networking', '37', '5', '27', '1']




**Now we're going to check the lenghts of the two free and non free datasets**

In [33]:
len_free_ios = len (ios_free)
print (len_free_ios)

len_non_free_ios = len (ios_non_free)
print (len_non_free_ios)

3253
2973


In [34]:
len_free_and = len (android_free)
print (len_free_and)

len_non_free_and = len (android_non_free)
print (len_non_free_and)

8878
754


**Everything seems fine so far.**

**To minimize risks on the development of apps, we have to have a validation strategy. We suggest the following steps:**

1. Build a minimal version of the app, only for Android, and add it to Google Play Store, since that platform costs for uploading an app is minimal, around US$ 25 one time payment.

2. We'll monitor the version performance amongst users. If the performance is good, we can develop the app further.

3. After six months, if the app shows profitable, we build and iOS version. This is necessary since there are bigger costs involved on uploading an app there.

**We'll now create a function to make a frequency table, where we can look up the sum of every occurence of a genre**

In [35]:
def freq_table (dataset, index, frequency=True):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        
        value = row[index]
        if value not in table:
            table[value] = 1
        if value in table:
            table[value] += 1
    
    percent_tab = {}
    for key in table:
        percentage = (float (table[key])/total)*100
        percent_tab[key] = percentage
        
    if frequency == True:
        return (table)
    if frequency == False:
        return (percent_tab)

**Below, we create another function to order the genres by their number of occurences in a descending order, from the largest to the smallest.**

In [36]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
    
    table_sorted = sorted (table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

**Now we use the functions to check the frequency of genres on the two datasets**

In [37]:
display_table(android_free, 1)

FAMILY : 1681
GAME : 864
TOOLS : 751
BUSINESS : 409
LIFESTYLE : 349
PRODUCTIVITY : 347
FINANCE : 329
MEDICAL : 314
SPORTS : 302
PERSONALIZATION : 296
COMMUNICATION : 288
HEALTH_AND_FITNESS : 274
PHOTOGRAPHY : 262
NEWS_AND_MAGAZINES : 251
SOCIAL : 237
TRAVEL_AND_LOCAL : 208
SHOPPING : 200
BOOKS_AND_REFERENCE : 193
DATING : 166
VIDEO_PLAYERS : 160
MAPS_AND_NAVIGATION : 126
FOOD_AND_DRINK : 111
EDUCATION : 104
ENTERTAINMENT : 86
LIBRARIES_AND_DEMO : 84
AUTO_AND_VEHICLES : 83
HOUSE_AND_HOME : 74
WEATHER : 72
EVENTS : 64
PARENTING : 59
ART_AND_DESIGN : 57
COMICS : 56
BEAUTY : 54


In [38]:
android_genre_qt = display_table(android_free, 9)

Tools : 750
Entertainment : 541
Education : 476
Business : 409
Lifestyle : 348
Productivity : 347
Finance : 329
Medical : 314
Sports : 308
Personalization : 296
Communication : 288
Action : 276
Health & Fitness : 274
Photography : 262
News & Magazines : 251
Social : 237
Travel & Local : 207
Shopping : 200
Books & Reference : 193
Simulation : 182
Dating : 166
Arcade : 165
Video Players & Editors : 158
Casual : 157
Maps & Navigation : 126
Food & Drink : 111
Puzzle : 101
Racing : 89
Role Playing : 84
Libraries & Demo : 84
Auto & Vehicles : 83
Strategy : 82
House & Home : 74
Weather : 72
Events : 64
Adventure : 62
Comics : 55
Beauty : 54
Art & Design : 53
Parenting : 45
Card : 41
Casino : 39
Trivia : 38
Educational;Education : 36
Board : 35
Educational : 34
Education;Education : 32
Word : 24
Casual;Pretend Play : 22
Music : 19
Racing;Action & Adventure : 16
Puzzle;Brain Games : 16
Entertainment;Music & Video : 16
Casual;Brain Games : 13
Casual;Action & Adventure : 13
Arcade;Action & Advent

In [39]:
ios_genre_qt = display_table (ios_free, 11)

Games : 1890
Entertainment : 262
Photo & Video : 163
Education : 119
Social Networking : 105
Shopping : 86
Utilities : 82
Sports : 70
Music : 67
Health & Fitness : 66
Productivity : 58
Lifestyle : 52
News : 45
Travel : 41
Finance : 37
Food & Drink : 31
Weather : 29
Reference : 19
Business : 18
Book : 16
Medical : 8
Navigation : 7
Catalogs : 5


**As we can see, "Games", "Entertainment", "Education", "Family" and "Tools" categories are the ones with more apps on the two datasets. But what are the most downloaded ones, proportionally? The average installation number by genre could show us this.**

**Android apps list has an "installs" column, we could use it for our analysis. But Apple apps list has no similar field, which leaves us with no choice besides using the "rating_count_tot" for measuring the number of times an app has been used and therefore rated by the users.**

**Now we're going to explore the Android installs column. But first, it need some cleansing on its values, since there are many of them presented with commas, plus simbols and others, as we can see below. The "installs" number can be seen on index 5**

In [40]:
explore_data(android_free, 0, 5)

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


['Smoke Effect Photo Maker - Smoke Editor', 'ART_AND_DESIGN', '3.8', '178', '19M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'April 26, 2018', '1.1', '4.0.3 and up']




In [41]:
android_installs_genre = {}

for row in android_free:
    genre = row[9]
    
    installs = row[5]
    installs = installs.replace(',','')
    installs = installs.replace('+','')
    installs = float(installs)
    
    if genre not in android_installs_genre:
        android_installs_genre[genre] = installs
    
    if genre in android_installs_genre:
        android_installs_genre[genre] += installs

**Now we have the number of times each genre was downloaded. With the function "sorted', let's check the occurences on new dictionary we created and the greatest values there.**

In [42]:
sorted(android_installs_genre.items(), key=lambda x: (-x[1], x[0]))

[('Communication', 12036906201.0),
 ('Tools', 9091043474.0),
 ('Social', 6487861902.0),
 ('Productivity', 6291679314.0),
 ('Video Players & Editors', 4916731720.0),
 ('Photography', 4657268815.0),
 ('Arcade', 3853691940.0),
 ('Action', 3470986940.0),
 ('Casual', 3057798570.0),
 ('Entertainment', 3015322513.0),
 ('Travel & Local', 2944604086.0),
 ('News & Magazines', 2378207260.0),
 ('Books & Reference', 1665985260.0),
 ('Personalization', 1579235988.0),
 ('Sports', 1421230683.0),
 ('Shopping', 1410338585.0),
 ('Racing', 1410136820.0),
 ('Health & Fitness', 1144048402.0),
 ('Strategy', 917192105.0),
 ('Puzzle', 880286191.0),
 ('Business', 706952090.0),
 ('Simulation', 679062620.0),
 ('Maps & Navigation', 603160780.0),
 ('Lifestyle', 497584429.0),
 ('Finance', 465163132.0),
 ('Weather', 410288520.0),
 ('Adventure', 395417120.0),
 ('Role Playing', 334148570.0),
 ('Education', 261887900.0),
 ('Casual;Action & Adventure', 255000000.0),
 ('Trivia', 228601370.0),
 ('Music', 220020500.0),
 ('W

**Looks like Communication apps are the most downloaded and installed ones, if we consider the "genres" column. But since Android apps dataset has seemly 2 columns for genre (genre itself and category), we have to check the other column.**

In [43]:
android_category_installs = {}

for row in android_free:
    category = row[1]
    
    installs = row[5]
    installs = installs.replace('+','')
    installs = installs.replace(',','')
    installs = float(installs)
    
    if category not in android_category_installs:
        android_category_installs[category] = installs
        
    if category in android_category_installs:
        android_category_installs[category] += installs

In [44]:
sorted(android_category_installs.items(), key=lambda x: (-x[1], x[0]))

[('GAME', 13446919450.0),
 ('COMMUNICATION', 12036906201.0),
 ('TOOLS', 9101043474.0),
 ('SOCIAL', 6487861902.0),
 ('PRODUCTIVITY', 6291679314.0),
 ('FAMILY', 6200925690.0),
 ('VIDEO_PLAYERS', 4931731720.0),
 ('PHOTOGRAPHY', 4657268815.0),
 ('TRAVEL_AND_LOCAL', 2944704086.0),
 ('NEWS_AND_MAGAZINES', 2378207260.0),
 ('BOOKS_AND_REFERENCE', 1665985260.0),
 ('PERSONALIZATION', 1579235988.0),
 ('SHOPPING', 1410338585.0),
 ('HEALTH_AND_FITNESS', 1144048402.0),
 ('SPORTS', 1095330683.0),
 ('ENTERTAINMENT', 990460000.0),
 ('BUSINESS', 706952090.0),
 ('MAPS_AND_NAVIGATION', 603160780.0),
 ('LIFESTYLE', 507584429.0),
 ('FINANCE', 465163132.0),
 ('WEATHER', 410288520.0),
 ('FOOD_AND_DRINK', 211788751.0),
 ('EDUCATION', 188950000.0),
 ('DATING', 150914757.0),
 ('ART_AND_DESIGN', 118211100.0),
 ('HOUSE_AND_HOME', 107202461.0),
 ('AUTO_AND_VEHICLES', 53180061.0),
 ('LIBRARIES_AND_DEMO', 53045810.0),
 ('COMICS', 45471150.0),
 ('MEDICAL', 38732344.0),
 ('PARENTING', 31481010.0),
 ('BEAUTY', 27697050.

**Comparing the two dictionaries we have just created, the values on the main categories/genres are exactly the same. The only exception found was on "Games" category, which was not on "genres" column, but appeared as the largest volume on the two columns, a fact we can not ignore.** 

**So we decided to keep the values on category column, which were just equal genres column, but more complete.**

**Now Let's divide this by the number of apps by genre, and check the average download per genre**

In [45]:
android_freq_gen = freq_table(android_free, 1)
sorted(android_freq_gen.items(), key=lambda x: (-x[1], x[0]))

[('FAMILY', 1681),
 ('GAME', 864),
 ('TOOLS', 751),
 ('BUSINESS', 409),
 ('LIFESTYLE', 349),
 ('PRODUCTIVITY', 347),
 ('FINANCE', 329),
 ('MEDICAL', 314),
 ('SPORTS', 302),
 ('PERSONALIZATION', 296),
 ('COMMUNICATION', 288),
 ('HEALTH_AND_FITNESS', 274),
 ('PHOTOGRAPHY', 262),
 ('NEWS_AND_MAGAZINES', 251),
 ('SOCIAL', 237),
 ('TRAVEL_AND_LOCAL', 208),
 ('SHOPPING', 200),
 ('BOOKS_AND_REFERENCE', 193),
 ('DATING', 166),
 ('VIDEO_PLAYERS', 160),
 ('MAPS_AND_NAVIGATION', 126),
 ('FOOD_AND_DRINK', 111),
 ('EDUCATION', 104),
 ('ENTERTAINMENT', 86),
 ('LIBRARIES_AND_DEMO', 84),
 ('AUTO_AND_VEHICLES', 83),
 ('HOUSE_AND_HOME', 74),
 ('WEATHER', 72),
 ('EVENTS', 64),
 ('PARENTING', 59),
 ('ART_AND_DESIGN', 57),
 ('COMICS', 56),
 ('BEAUTY', 54)]

In [46]:
avg_inst_gen = {k:v1/android_freq_gen.get(k,0) for k,v1 in android_category_installs.items()}

In [47]:
sorted(avg_inst_gen.items(), key=lambda x: (-x[1], x[0]))

[('COMMUNICATION', 41794813.197916664),
 ('VIDEO_PLAYERS', 30823323.25),
 ('SOCIAL', 27374944.734177217),
 ('PRODUCTIVITY', 18131640.674351584),
 ('PHOTOGRAPHY', 17775835.171755724),
 ('GAME', 15563564.17824074),
 ('TRAVEL_AND_LOCAL', 14157231.182692308),
 ('TOOLS', 12118566.543275632),
 ('ENTERTAINMENT', 11516976.744186046),
 ('NEWS_AND_MAGAZINES', 9474929.322709164),
 ('BOOKS_AND_REFERENCE', 8632047.979274612),
 ('SHOPPING', 7051692.925),
 ('WEATHER', 5698451.666666667),
 ('PERSONALIZATION', 5335256.716216216),
 ('MAPS_AND_NAVIGATION', 4786990.317460317),
 ('HEALTH_AND_FITNESS', 4175359.131386861),
 ('FAMILY', 3688831.463414634),
 ('SPORTS', 3626922.7913907287),
 ('ART_AND_DESIGN', 2073878.9473684211),
 ('FOOD_AND_DRINK', 1908006.7657657657),
 ('EDUCATION', 1816826.923076923),
 ('BUSINESS', 1728489.2176039119),
 ('LIFESTYLE', 1454396.6446991405),
 ('HOUSE_AND_HOME', 1448681.9054054054),
 ('FINANCE', 1413869.7021276595),
 ('DATING', 909125.0421686746),
 ('COMICS', 811984.8214285715),


**As we can see, Android apps set is dominated by gaming and communication apps. Let's know further the top apps on the top genres, starting with the "Communication" ones.**

In [62]:
communication_apps = []

for row in android_free:
    genre = row[1]
    if genre == 'COMMUNICATION':
        communication_apps.append(row)

In [65]:
sorted (communication_apps)

[['/u/app',
  'COMMUNICATION',
  '4.7',
  '573',
  '53M',
  '10,000+',
  'Free',
  '0',
  'Mature 17+',
  'Communication',
  'July 3, 2018',
  '4.2.4',
  '4.1 and up'],
 ['2ndLine - Second Phone Number',
  'COMMUNICATION',
  '4.2',
  '15880',
  '32M',
  '1,000,000+',
  'Free',
  '0',
  'Everyone',
  'Communication',
  'August 3, 2018',
  '5.65.0',
  '4.0 and up'],
 ['3G DZ Configuration',
  'COMMUNICATION',
  '4.3',
  '488',
  '2.5M',
  '50,000+',
  'Free',
  '0',
  'Everyone',
  'Communication',
  'November 19, 2015',
  '6.2.3',
  '3.0 and up'],
 ['AC-BL',
  'COMMUNICATION',
  'NaN',
  '0',
  '4.3M',
  '50+',
  'Free',
  '0',
  'Everyone',
  'Communication',
  'April 20, 2018',
  '1.7',
  '2.1 and up'],
 ['AG Contacts, Lite edition',
  'COMMUNICATION',
  '4.5',
  '185',
  '3.7M',
  '5,000+',
  'Free',
  '0',
  'Everyone',
  'Communication',
  'May 18, 2018',
  '8.05.18.14250',
  '4.0.3 and up'],
 ['AK Phone',
  'COMMUNICATION',
  '4.4',
  '9',
  '4.2M',
  '5,000+',
  'Free',
  '0',
  

**As we can see, apps like Skype, Google Chrome, WhatsApp and AT&T Messenger fall into the "Communication". Let's check other categories, starting by the second on average installs: "Video Players".**

In [66]:
video_players_apps = []

for row in android_free:
    genre = row[1]
    if genre == 'VIDEO_PLAYERS':
        video_players_apps.append(row)

In [68]:
sorted (video_players_apps)

[['4K VIDEO PLAYER ULTRA HD',
  'VIDEO_PLAYERS',
  '3.7',
  '20',
  '24M',
  '5,000+',
  'Free',
  '0',
  'Everyone',
  'Video Players & Editors',
  'May 1, 2017',
  '7.0',
  '2.3.3 and up'],
 ['A-B repeater',
  'VIDEO_PLAYERS',
  '4.4',
  '32',
  '239k',
  '5,000+',
  'Free',
  '0',
  'Everyone',
  'Video Players & Editors',
  'May 18, 2018',
  '1.8',
  '3.0 and up'],
 ['A-Z Screen Recorder -',
  'VIDEO_PLAYERS',
  'NaN',
  '17',
  '3.7M',
  '500+',
  'Free',
  '0',
  'Everyone',
  'Video Players & Editors',
  'July 23, 2018',
  '1.0',
  '4.4 and up'],
 ['AB Repeat Player',
  'VIDEO_PLAYERS',
  '4.3',
  '1628',
  '18M',
  '100,000+',
  'Free',
  '0',
  'Everyone',
  'Video Players & Editors',
  'March 6, 2018',
  '4.1.1',
  '4.4 and up'],
 ['ACTIVEON CX & CX GOLD',
  'VIDEO_PLAYERS',
  '3.3',
  '439',
  '35M',
  '50,000+',
  'Free',
  '0',
  'Everyone',
  'Video Players & Editors',
  'March 4, 2016',
  '0.7.9',
  '4.0 and up'],
 ['AJ Player',
  'VIDEO_PLAYERS',
  '4.8',
  '18',
  '1.7

**Video players category has an interesting panorama, including video capturing solutions, players and editors. Now with the third place, Social.**

In [80]:
social_apps = []

for row in android_free:
    genre = row[1]
    if genre == 'SOCIAL':
        social_apps.append(row)

In [81]:
sorted (social_apps)

[['+Download 4 Instagram Twitter',
  'SOCIAL',
  '4.5',
  '40467',
  '22M',
  '1,000,000+',
  'Free',
  '0',
  'Everyone',
  'Social',
  'August 2, 2018',
  '5.03',
  '4.1 and up'],
 ['Alarm.fo – choose your info',
  'SOCIAL',
  '4.4',
  '19',
  '3.1M',
  '1,000+',
  'Free',
  '0',
  'Everyone',
  'Social',
  'May 29, 2018',
  '1.4.8',
  '4.1 and up'],
 ['All Social Networks',
  'SOCIAL',
  '4.2',
  '22650',
  '1.5M',
  '1,000,000+',
  'Free',
  '0',
  'Everyone',
  'Social',
  'May 21, 2018',
  '2.4.12',
  '4.0 and up'],
 ['Alumni BJ',
  'SOCIAL',
  'NaN',
  '2',
  '5.4M',
  '100+',
  'Free',
  '0',
  'Everyone',
  'Social',
  'November 8, 2017',
  '1.0',
  '4.1 and up'],
 ['Amino: Communities and Chats',
  'SOCIAL',
  '4.8',
  '1264084',
  '62M',
  '10,000,000+',
  'Free',
  '0',
  'Teen',
  'Social',
  'August 7, 2018',
  '1.8.19179',
  '4.0.3 and up'],
 ['Amleen Ey',
  'SOCIAL',
  'NaN',
  '0',
  '19M',
  '1+',
  'Free',
  '0',
  'Everyone',
  'Social',
  'July 24, 2017',
  '1.0',


**Social apps category includes all the big social media apps, like Facebook, Twitter, Instagram and tools for a better use of those.**

**Let' check the apps competition on these three categories.**

In [79]:
comlen = len (communication_apps)
vilen = len (video_players_apps)
soclen = len (social_apps)

print (comlen)
print (vilen)
print (soclen)

287
159
236


**With 287 entries, the category with more apps competing for user visibility is "Communication". Besides the competition, this genre also has some big and well-stablished players, and by elimination, we concluded we should look somewhere else for better opportunities.**

**The "Social" apps category is the second more competitive. We found some big players there too.**

**By elimination, we indicate the "video players" category as the best opportunity for developing an app.**